File: servers/server-tcp.lua
local socket = require("socket")

print("Starting TCP demo-server on port 12345")

local tcp = socket.bind('127.0.0.1', 12345)

local running = true
while running do
  local client = tcp:accept()
  client:settimeout(2)
  local data, err = client:receive()
  if data then
    print("OK: " .. data)
    client:send(data .. "\n")
  else
    print("Error: " .. err)
  end
  client:close()
  socket.sleep(0.01)
end

File: servers/server-tcp.py
# this is a simple TCP socket echo-server (send back upper-case)

import socketserver

class MyTCPHandler(socketserver.BaseRequestHandler):
    def handle(self):
        self.data = self.request.recv(1024).strip()
        print(f"{self.client_address[0]} wrote: {self.data}")
        self.request.sendall(self.data.upper())

print("Starting simple TCP socket server on port 12345")
with socketserver.TCPServer(('0.0.0.0', 12345), MyTCPHandler) as server:
    server.serve_forever()